R/tokenizers.R

```diff
@@ -228,3 +229,71 @@ normalize_characters <- function(x) {
 
     return(x)
 }
+
+#' @rdname tokenize_internal
+#' @export
+tokenize_word4 <- function(x, split_hyphens = FALSE, split_tags = FALSE, split_elisions = FALSE,
+                           verbose = quanteda_options("verbose"), ...) {
+    
+    rules <- quanteda::data_breakrules_word
+    if (!split_hyphens) {
+        if (verbose) catm(" ...preserving hyphens\n")
+    } else {
+        rules[["keep_hyphens"]] <- NULL
+    }
+    if (!split_elisions) {
+        if (verbose) catm(" ...preserving elisions\n")
+        rules[["split_elisions"]] <- NULL
+    }
+    if (!split_tags) {
+        if (verbose) catm(" ...preserving social media tags (#, @)\n")
+        # NOTE: default patterns are protected by keep_tags
+        rules[["keep_usernames"]] <- paste0(stri_replace_all_fixed(quanteda_options("pattern_username"), "@", "\\@"), ";")
+        rules[["keep_hashtags"]] <- paste0(stri_replace_all_fixed(quanteda_options("pattern_hashtag"), "#", "\\#"), ";")
+    } else {
+        rules[["keep_tags"]] <- NULL
+    }
+    tokenize_custom(x, rules)
+}
+
+#' Customizable tokenizer
+#'
+#' Allows users to tokenize texts using customized boundary rules. See the [ICU
+#' website](https://unicode-org.github.io/icu/userguide/boundaryanalysis/break-rules.html)
+#' for how to define boundary rules.
+#' @param x character vector for texts to tokenize
+#' @param rules a list of rules for rule-based boundary detection
+#' @importFrom stringi stri_split_boundaries
```

Comment:
I'm afraid we need to get into the habit now of including @return values on all (new) documented functions. (See https://r-pkgs.org/man.html#sec-man-returns) I just spent a full day fixing this on **quanteda.textmodels**.

Reply:
That is the right thing to do.

Reply:
Done